### PR TITLE
Refactor of smartos_vmadm module

### PR DESCRIPTION
**Pull request by sjorge**
This is part 1 of 2 (or 3) of the vmadm execution module refactor mentioned in #26507 
- moved smartos_vmadm.py to smartos_legacy.py
- added deprecation warning to smartos_legacy.py (**virtual** = 'virt')
- create new smartos_vmadm.py (**virtual** = 'vmadm') that closely mirrors the vmadm command on SmartOS
## vmadm.create

```
[root@core ~]# /opt/local/salt/bin/salt-call --local vmadm.create brand=joyent image_uuid=c02a2044-c1bd-11e4-bd8c-dfc1db8b0182 hostname=test.example.org alias=test autoboot=false nowait=false cpu_shares=50 cpu_cap=125 max_physical_memory=256 quota=1 zfs_io_priority=50 zfs_root_compression=lz4 nics='[{ "nic_tag": "trunk", "mtu": 1500, "vlan_id": 10, "ip": "172.16.xx.121", "netmask": "255.255.255.0" }]'
[INFO    ] Executing command 'echo \'{"nowait": false, "zfs_root_compression": "lz4", "alias": "test", "nics": [{"ip": "172.16.xx.121", "netmask": "255.255.255.0", "nic_tag": "trunk", "vlan_id": 10, "mtu": 1500}], "brand": "joyent", "hostname": "test.example.org", "quota": 1, "max_physical_memory": 256, "autoboot": false, "cpu_cap": 125, "image_uuid": "c02a2044-c1bd-11e4-bd8c-dfc1db8b0182", "zfs_io_priority": 50, "cpu_shares": 50}\' | /usr/sbin/vmadm validate create ' in directory '/root'
[INFO    ] Executing command 'echo \'{"nowait": false, "zfs_root_compression": "lz4", "alias": "test", "nics": [{"ip": "172.16.xx.121", "netmask": "255.255.255.0", "nic_tag": "trunk", "vlan_id": 10, "mtu": 1500}], "brand": "joyent", "hostname": "test.example.org", "quota": 1, "max_physical_memory": 256, "autoboot": false, "cpu_cap": 125, "image_uuid": "c02a2044-c1bd-11e4-bd8c-dfc1db8b0182", "zfs_io_priority": 50, "cpu_shares": 50}\' | /usr/sbin/vmadm create ' in directory '/root'
local:
    4ed3416e-73bf-4f87-b49c-3b33eb6b520d
```
## vmadm.update

```
[root@core ~]# /opt/local/salt/bin/salt-call --local vmadm.update vm=4ed3416e-73bf-4f87-b49c-3b33eb6b520d max_physical_memory=1024
[INFO    ] Executing command '/usr/sbin/vmadm lookup -1  uuid=4ed3416e-73bf-4f87-b49c-3b33eb6b520d' in directory '/root'
[INFO    ] Executing command '/usr/sbin/vmadm lookup -1  uuid=4ed3416e-73bf-4f87-b49c-3b33eb6b520d' in directory '/root'
[INFO    ] Executing command '/usr/sbin/vmadm get 4ed3416e-73bf-4f87-b49c-3b33eb6b520d' in directory '/root'
[INFO    ] Executing command 'echo \'{"max_physical_memory": 1024}\' | /usr/sbin/vmadm validate update joyent' in directory '/root'
[INFO    ] Executing command 'echo \'{"max_physical_memory": 1024}\' | /usr/sbin/vmadm update 4ed3416e-73bf-4f87-b49c-3b33eb6b520d' in directory '/root'
local:
    True
```
## vmadm.get

```
[root@core ~]# /opt/local/salt/bin/salt-call --local vmadm.get 4ed3416e-73bf-4f87-b49c-3b33eb6b520d
[INFO    ] Executing command '/usr/sbin/vmadm lookup -1  uuid=4ed3416e-73bf-4f87-b49c-3b33eb6b520d' in directory '/root'
[INFO    ] Executing command '/usr/sbin/vmadm get 4ed3416e-73bf-4f87-b49c-3b33eb6b520d' in directory '/root'
local:
    ----------
    alias:
        test
    autoboot:
        False
    billing_id:
        00000000-0000-0000-0000-000000000000
    brand:
        joyent
    cpu_cap:
        125
    cpu_shares:
        50
    create_timestamp:
        2015-08-28T14:34:27.951Z
    customer_metadata:
        ----------
    dns_domain:
        local
    firewall_enabled:
        False
    hostname:
        test.example.org
    image_uuid:
        c02a2044-c1bd-11e4-bd8c-dfc1db8b0182
    internal_metadata:
        ----------
    last_modified:
        2015-08-28T14:36:30.000Z
    limit_priv:
        default
    max_locked_memory:
        1024
    max_lwps:
        2000
    max_msg_ids:
        4096
    max_physical_memory:
        1024
    max_sem_ids:
        4096
    max_shm_ids:
        4096
    max_shm_memory:
        256
    max_swap:
        1024
    nics:
        |_
          ----------
          interface:
              net0
          ip:
              172.16.xx.121
          mac:
              42:6c:da:cb:97:af
          mtu:
              1500
          netmask:
              255.255.255.0
          nic_tag:
              trunk
          primary:
              True
          vlan_id:
              10
    owner_uuid:
        00000000-0000-0000-0000-000000000000
    platform_buildstamp:
        20150820T062742Z
    quota:
        1
    resolvers:
    routes:
        ----------
    server_uuid:
        00000000-0000-0000-0000-002590f15506
    snapshots:
    state:
        stopped
    tags:
        ----------
    tmpfs:
        1024
    uuid:
        4ed3416e-73bf-4f87-b49c-3b33eb6b520d
    v:
        1
    zfs_filesystem:
        zones/4ed3416e-73bf-4f87-b49c-3b33eb6b520d
    zfs_io_priority:
        50
    zfs_root_compression:
        lz4
    zfs_root_recsize:
        131072
    zone_state:
        installed
    zoneid:
        None
    zonename:
        4ed3416e-73bf-4f87-b49c-3b33eb6b520d
    zonepath:
        /zones/4ed3416e-73bf-4f87-b49c-3b33eb6b520d
    zpool:
        zones
```
## vmadm.lookup

```
[root@core ~]# /opt/local/salt/bin/salt-call --local vmadm.lookup search='alias=opnsense' one=True
[INFO    ] Executing command '/usr/sbin/vmadm lookup -1  alias=opnsense' in directory '/root'
local:
    186da9ab-7392-4f55-91a5-b8f1fe770543
```
## vmadm.start

```
[root@core ~]# /opt/local/salt/bin/salt-call --local vmadm.start 4ed3416e-73bf-4f87-b49c-3b33eb6b520d
[INFO    ] Executing command '/usr/sbin/vmadm lookup -1  uuid=4ed3416e-73bf-4f87-b49c-3b33eb6b520d' in directory '/root'
[INFO    ] Executing command '/usr/sbin/vmadm start 4ed3416e-73bf-4f87-b49c-3b33eb6b520d ' in directory '/root'
local:
    True
```
## vmadm.stop

```
[root@core ~]# /opt/local/salt/bin/salt-call --local vmadm.stop 4ed3416e-73bf-4f87-b49c-3b33eb6b520d force=True
[INFO    ] Executing command '/usr/sbin/vmadm lookup -1  uuid=4ed3416e-73bf-4f87-b49c-3b33eb6b520d' in directory '/root'
[INFO    ] Executing command '/usr/sbin/vmadm stop -F 4ed3416e-73bf-4f87-b49c-3b33eb6b520d' in directory '/root'
local:
    True
```
## vmadm.info

```
[root@core ~]# /opt/local/salt/bin/salt-call --local vmadm.info opnsense vnc key=alias
[INFO    ] Executing command '/usr/sbin/vmadm lookup -1  alias=opnsense' in directory '/root'
[INFO    ] Executing command '/usr/sbin/vmadm info 186da9ab-7392-4f55-91a5-b8f1fe770543 vnc' in directory '/root'
local:
    ----------
    vnc:
        ----------
        display:
            1
        host:
            172.16.xx.120
        port:
            5901
```
## vmadm.list

```
[root@core ~]# /opt/local/salt/bin/salt-call --local vmadm.list
[INFO    ] Executing command '/usr/sbin/vmadm list -p -H -o uuid,type,ram,state,alias  ' in directory '/root'
local:
    |_
      ----------
      alias:
          nina
      ram:
          2048
      state:
          stopped
      type:
          KVM
      uuid:
          618a01b8-c37b-473f-8cdb-ca2a66d6adf7
    |_
      ----------
      alias:
          jiska
      ram:
          4096
      state:
          running
      type:
          OS
      uuid:
          260014e8-33db-42c0-8ce4-2a7929a5ef08
    |_
      ----------
      alias:
          silke
      ram:
          8192
      state:
          running
      type:
          LX
      uuid:
          306b5291-3dc6-4947-8faa-6ec8da48ca6d
```

... there are a few more commands but this should give the jist of it. Most command vmadm has are present with the exaction of vmadm console which cannot be implemented without and interactive TTY.

I will be on holiday next week, after that I will write a simplified wrapper to replace smartos_legacy.py and hopefully a state module when I figure out how they work.

### CODE REVIEW

File: salt/modules/smartos_legacy.py [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+
+
+def __virtual__():
+    '''
+    Provides virt on SmartOS
+    '''
+    if salt.utils.is_smartos_globalzone() and _check_vmadm():
+        # Warn about deprication
+        salt.utils.warn_until(
+            'Aluminum',
```
Comment (jfindlay) [inline salt/modules/smartos_legacy.py]: `Nitrogen` is probably far enough in the future for this change.

**Comment by jfindlay** [comment]
This is excellent, @sjorge.  We are working on stabilizing 2015.8.0 for release and this is a large change.  I don't know of any SmartOS users besides you and @nshalman, so I am unsure what the impact would be.  Since it is only confined to SmartOS it may be fine.  I will defer the decision to merge this into 2015.8 to @thatch45 and @meggiebot.

**Comment by sjorge** [comment]
I will make the change to nitrogen once I get home.

Impact shouldn't be too high. I chose to keep the original module around for compatibility so nothing should break on that front.

**Comment by basepi** [comment]
Cool. Yeah, I'm happy to merge this once the warn_until is updated.

**Comment by sjorge** [comment]
Change pushed, so once I get back from my holiday and I start working on the state and wrapper modules... which release do I target?

**Comment by basepi** [comment]
Great! If you do the same deprecation path for the state/wrapper modules, you can target 2015.8 or develop, either way. Technically it should be develop because it's new functionality, but if it doesn't break anything existing we can potentially sneak it in.

**Comment by sjorge** [comment]
There currently is no state module :) so it will be new stuff, so the chances if it breaking things should be small.

**Comment by basepi** [comment]
Sounds god.

**Comment by basepi** [comment]
Hey @sjorge, this is conflicting in huge ways with the work that @isbm has done on the develop branch. 17 conflicting sections. Because I do not know the code, I am not equipped to handle this merge conflict. So, for the moment, I am going to take the develop branch's version.

I know it's far from convenient, but could you make your additions separately on that branch? That's the best way I can see for combining yours and @isbm's changes, is if you make yours on top of his.

Sorry again for the inconvenience.

**Comment by sjorge** [comment]
Alright, i'll try to get that done before i hop on the plain next week.

Looks simple enough, move develops smartos_vmadm.py to smartos_legacy.py and drop in the new smartos_vmadm.py.

**Comment by basepi** [comment]
That seems reasonable to me. You might ping @isbm on the pull request so he can give feedback. Might have some good additions to your new code.

**Comment by sjorge** [comment]
The code cleanup done by @isbm looks pretty good. That more or less takes care of writing new 'virt' wrapper for smartos.

Will fix it up so both the new 'vmadm' and the fixed up 'virt' will be around. Makes sense to bring it over to 2015.8 also. Will do 2 PR's somewhere this weekend.

**Comment by basepi** [comment]
Awesome! Thanks so much @sjorge!

**Comment by sjorge** [comment]
PR #26772 and #26770 together bring develop and 2015.8 in sync in regards to the smartos 'virt', 'imgadm' and 'vmadm' modules.

**Comment by basepi** [comment]
Thank you @sjorge!

**Pull request merged**